### PR TITLE
Add SMT Writer code to populate SMT.

### PR DESCRIPTION
**Pull request by AlCutter**
Not complete yet: missing code to pull in previous nodes; a bit go-routine happy; etc. but works for calculating SMT roots from scratch.

### CODE REVIEW

File: merkle/sparse_merkle_tree.go [hunk 1]
```diff
@@ -3,9 +3,11 @@ package merkle
 import (
 	"errors"
 	"fmt"
+	"math/big"
 
 	"github.com/google/trillian"
 	"github.com/google/trillian/storage"
+	"sync"
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree.go]: ordering

File: merkle/sparse_merkle_tree.go [hunk 2]
```diff
@@ -25,7 +27,191 @@ type SparseMerkleTreeReader struct {
...
+// Subtree is an interface which must be implemented by subtree workers.
+// Currently there's only a locally sharded go-routine based implementation,
+// the the idea is that an RPC based sharding implementation could be created
+// and dropped in.
+type Subtree interface {
+	// SetLeaf sets a single leaf hash for integration into a sparse merkle tree.
+	SetLeaf(index []byte, hash trillian.Hash) error
+
+	// CalculateRoot instructs the subtree worker to start calculating the root
+	// has of its tree.  It is an error to call SetLeaf() after calling this
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree.go]: typo

File: merkle/sparse_merkle_tree.go [hunk 3]
```diff
@@ -25,7 +27,191 @@ type SparseMerkleTreeReader struct {
...
+	childPrefixStr := string(childPrefix)
+	s.childMutex.Lock()
+	defer s.childMutex.Unlock()
+
+	subtree := s.children[childPrefixStr]
+	if subtree == nil {
+		subtree = s.getSubtree(childPrefix)
+		s.children[childPrefixStr] = subtree
+
+		// Since a new subtree worker is being created, then we'll add a future is
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree.go]: Reads strangely, poss needs an edit.

File: merkle/sparse_merkle_tree.go [hunk 4]
```diff
@@ -45,13 +231,47 @@ func NewSparseMerkleTreeReader(rev int64, h MapHasher, tx storage.ReadOnlyTreeTX
 	}
 }
 
+func leafQueueSize(depths []int) int {
+	if len(depths) == 1 {
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree.go]: Would this go wrong if someone passed 0 or -1 for example?

File: merkle/sparse_merkle_tree.go [hunk 5]
```diff
@@ -45,13 +231,47 @@ func NewSparseMerkleTreeReader(rev int64, h MapHasher, tx storage.ReadOnlyTreeTX
...
+	// this, but it'll do for now.
+	go tree.buildSubtree()
+	return &tree
+}
+
 // NewSparseMerkleTreeWriter returns a new SparseMerkleTreeWriter, which will
 // write data back into the tree at the specified revision, using the passed
 // in MapHasher to calulate/verify tree hashes, storing via tx.
 func NewSparseMerkleTreeWriter(rev int64, h MapHasher, tx storage.TreeTX) *SparseMerkleTreeWriter {
+	topSubtreeSize := 8 // must be a multiple of 8 for now.
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree.go]: Might be worth pulling out the 8s as constant.

File: merkle/sparse_merkle_tree.go [hunk 6]
```diff
@@ -122,11 +342,20 @@ func (s SparseMerkleTreeReader) InclusionProof(rev int64, key trillian.Key) ([]t
 	return r, nil
 }
 
-// SetLeaves commits the tree to setting a batch leaves to the specified value.
-// There may be a delay between the transaction this request is part of
-// successfully commiting, and the updated values being reflected in the served tree.
-func (s *SparseMerkleTreeWriter) SetLeaves(newRevision int64, leaves []HashKeyValue) (trillian.TreeRoot, error) {
-	return trillian.TreeRoot{}, errors.New("unimplemented")
+// SetLeaves adds a batch of leaves to the in-flight tree update.
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree.go]: Spurious blank line after comment or GH displaying it it wrong?

**Comment by Martin2112** [comment]
Seems basically LGTM with a few suggestions.

File: merkle/sparse_merkle_tree.go (hunk 1 above)
Comment (AlCutter) [inline merkle/sparse_merkle_tree.go]: done

File: merkle/sparse_merkle_tree.go (hunk 2 above)
Comment (AlCutter) [inline merkle/sparse_merkle_tree.go]: fixored

File: merkle/sparse_merkle_tree.go (hunk 3 above)
Comment (AlCutter) [inline merkle/sparse_merkle_tree.go]: yeah, that was gibberish - I've changed it a bit and hopefully it's a bit clearer.

File: merkle/sparse_merkle_tree.go (hunk 4 above)
Comment (AlCutter) [inline merkle/sparse_merkle_tree.go]: Remarkably, and quite unexpectedly: no - I just tried it (!)

Comment (Martin2112) [inline merkle/sparse_merkle_tree.go]: Job's a good un then!

File: merkle/sparse_merkle_tree.go (hunk 5 above)
Comment (AlCutter) [inline merkle/sparse_merkle_tree.go]: I've made it a const, but at some point I'll probably open it up to be configurable somehow - I've popped in a `TODO` for that.

File: merkle/sparse_merkle_tree.go (hunk 6 above)
Comment (AlCutter) [inline merkle/sparse_merkle_tree.go]: Don't see it locally...

**Pull request merged**